### PR TITLE
Partial fix issue 19294 (Updated with broader support for array operations)

### PR DESCRIPTION
**Pull request by sorin-gabriel**
Binary expressions with arrays of user-defined types now have more support. Before, an expression like `T[] op U` or `T[] op U[]` would fail to compile, even if the user-defined types T and U implemented opBinary or opBinaryRight.

In the case of 3+ operands, `T[] op T op U` and `T op T[] op U` are still unsupported (see _test/runnable/b19294.d_).

### CODE REVIEW

File: test/compilable/b19294.d [hunk 1]
```diff
@@ -1,69 +1,180 @@
-alias MT = MyStruct!int;
+alias T = MyStruct!float;
```
Comment (RazvanN7) [inline test/compilable/b19294.d:1]: Why change this test and not add a different one?

File: test/compilable/b19294.d [hunk 2]
```diff
@@ -1,69 +1,180 @@
...
 
-struct MyStruct(T)
+struct MyStruct(U)
```
Comment (RazvanN7) [inline test/compilable/b19294.d:3]: This change is not required. Please try to keep the diff minimal.

File: src/dmd/dcast.d [hunk 3]
```diff
@@ -3827,3 +3835,58 @@ IntRange getIntRange(Expression e)
     e.accept(v);
     return v.range;
 }
+
+Type checkArrayOpType(Expression e1, Expression e2, TOK op, Scope *sc)
```
Comment (RazvanN7) [inline src/dmd/dcast.d:3839]: public functions need to be documented, however, it seems that this function could be private. Either way, please document it using the [ddoc style](https://dlang.org/spec/ddoc.html).

File: src/dmd/dcast.d [hunk 4]
```diff
@@ -3827,3 +3835,58 @@ IntRange getIntRange(Expression e)
...
+{
+    Expression lhs = e1;
```
Comment (RazvanN7) [inline src/dmd/dcast.d:3841]: why not name the parameters lhs and rhs? That way, these declarations become useless.

File: test/compilable/b19294.d [hunk 5]
```diff
@@ -1,69 +1,180 @@
-alias MT = MyStruct!int;
+alias T = MyStruct!float;
 
-struct MyStruct(T)
```
Comment (thewilsonator) [inline test/compilable/b19294.d:3]: Please leave this the same, it creates unnecessary diff.

File: src/dmd/dcast.d [hunk 6]
```diff
@@ -3827,3 +3835,58 @@ IntRange getIntRange(Expression e)
...
+
+    // scalar op scalar - we shouldn't be here
+    if (lhs.type.ty != Tarray && lhs.type.ty != Tsarray && rhs.type.ty != Tarray && rhs.type.ty != Tsarray)
+        return null;
+
+    if (!lhs.isSliceExp() && !lhs.isArrayLiteralExp() && !rhs.isSliceExp() && !rhs.isArrayLiteralExp())
+        return null;
+
+    // T[x .. y] op ?
+    if (e1.isSliceExp())
```
Comment (RazvanN7) [inline src/dmd/dcast.d:3852]: it's confusing to use both e1 and lhs. Choose one and stick with it.

File: test/compilable/b19294.d [hunk 7]
```diff
@@ -1,69 +1,180 @@
...
+    ref MyStruct opOpAssign(string op, R)(const R z)
+        if (is(R == int) || is(R == float))
     {
         mixin ("x "~op~"= z;");
         return this;
     }
 }
 
-void test()
+void main()
```
Comment (thewilsonator) [inline test/compilable/b19294.d:55]: if this is supposed to be a runnable test move the file to runnable, otherwise leave the name unchanged.

**Review by RazvanN7 — CHANGES_REQUESTED** [review]
Please explain why the code was wrong, what your fix is doing and also some code comments would not be bad.

**Review by thewilsonator — DISMISSED** [review]
Please remove the first seven commit with either a rebase and squash or just reconstitute the diff on `dlang/master`

File: src/dmd/dcast.d (hunk 6 above)
Comment (sorin-gabriel) [reply src/dmd/dcast.d]: At the end of the execution I want to restore e1 and e2. At least one of them will be changed inside the function. I rearranged the code a bit, I hope now it makes more sense.

**Comment by sorin-gabriel** [comment]
Left _test/compilable/b19294.d_ as is and added _test/runnable/b19294.d_

**Review by RazvanN7 — APPROVED** [review]

**Comment by WalterBright** [comment]
> Binary expressions with arrays of user-defined types now have more support.

What exactly does "have more support" mean? Are the language semantics changed? Does this need a spec PR?

**Comment by RazvanN7** [comment]
@WalterBright Before this patch, this worked:

```d
void main()
{
     int[] a =[1, 2, 3];
     int[] b;
     b[] = a[] * 3;
}
```

But this didn't:

```d
struct A
{
      int a;
      A opBinary(string op)(int b) { mixin("return A(a" ~ op ~ " b);");}
}

void main()
{
     A[] a = [A(), A(), A()];
     A[] b;
     b[] = a[] * 3;            // incompatible types A[] and int
}
```

This patch makes it so that the latter works. This should not require a language spec as the expectation is that the the compiler will try to do vector operations regardless of the type of the array. Of course, another solution would be to simply modify the spec and disallow vector operations for arrays for which the basetype is not a basic type, but I don't see any reason why we should not support this. The compiler lowers such type of operations druntime calls anyway.

**Comment by WalterBright** [comment]
@RazvanN7 thank you for the explanation. What about unary operators?

I checked the coverage report, and there are a lot of cases untested. They should be fairly straightforward to test.

**Comment by RazvanN7** [comment]
Well, the [issue](https://issues.dlang.org/show_bug.cgi?id=19294) that this is trying to fix is specifically for binary operators. I, personally, did not even know until today that it is possible to do this:

```d
void main()
{
    int[] a = [1, 2, 3];
    ++a[];
}
```

That's pretty neat.

However, I think that unary operators could be tackled in a different PR.

**Comment by RazvanN7** [comment]
@WalterBright ok to merge this?
